primitives/header/src/lib.rs

```diff
@@ -0,0 +1,327 @@
+{
+	let u256: U256 = (*val).into();
+	serde::Serialize::serialize(&u256, s)
+}
+
+#[cfg(feature = "serde")]
+pub fn deserialize_number<'a, D, T: Copy + Into<U256> + TryFrom<U256>>(d: D) -> Result<T, D::Error>
+where
+	D: serde::Deserializer<'a>,
+{
+	let u256: U256 = serde::Deserialize::deserialize(d)?;
+	TryFrom::try_from(u256).map_err(|_| serde::de::Error::custom("Try from failed"))
+}
+
+impl<Number, Hash> sp_runtime::traits::Header for Header<Number, Hash>
+where
+	Number: BlockNumber,
+	Hash: HashT,
+	Hash::Output: From<[u8; 32]>,
+{
+	type Number = Number;
+	type Hash = <Hash as HashT>::Output;
+	type Hashing = Hash;
+
+	fn new(
+		number: Self::Number,
+		extrinsics_root: Self::Hash,
+		state_root: Self::Hash,
+		parent_hash: Self::Hash,
+		digest: Digest,
+	) -> Self {
+		Self { number, extrinsics_root, state_root, parent_hash, digest }
+	}
+	fn number(&self) -> &Self::Number {
+		&self.number
+	}
+
+	fn set_number(&mut self, num: Self::Number) {
+		self.number = num
+	}
+	fn extrinsics_root(&self) -> &Self::Hash {
+		&self.extrinsics_root
+	}
+
+	fn set_extrinsics_root(&mut self, root: Self::Hash) {
+		self.extrinsics_root = root
+	}
+	fn state_root(&self) -> &Self::Hash {
+		&self.state_root
+	}
+
+	fn set_state_root(&mut self, root: Self::Hash) {
+		self.state_root = root
+	}
+	fn parent_hash(&self) -> &Self::Hash {
+		&self.parent_hash
+	}
+
+	fn set_parent_hash(&mut self, hash: Self::Hash) {
+		self.parent_hash = hash
+	}
+
+	fn digest(&self) -> &Digest {
+		&self.digest
+	}
+
+	fn digest_mut(&mut self) -> &mut Digest {
+		#[cfg(feature = "std")]
+		log::debug!(target: "header", "Retrieving mutable reference to digest");
+		&mut self.digest
+	}
+	// We override the default hashing function to use
+	// a felt aligned pre-image for poseidon hashing.
+	fn hash(&self) -> Self::Hash {
+		let max_encoded_felts = 4 * 3 + 1 + 28; // 3 hashout fields + 1 u32 + 28 felts for injective digest encoding
+		let mut felts = Vec::with_capacity(max_encoded_felts);
+
+		// parent_hash : 32 bytes → 4 felts
+		felts.extend(unsafe_digest_bytes_to_felts::<Goldilocks>(
+			self.parent_hash.as_ref().try_into().expect("hash is 32 bytes"),
+		));
+
+		// block number as u64 (compact encoded, but we only need the value)
+		// constrain the block number to be with u32 range for simplicity
+		let number = self.number.into();
+		felts.push(Goldilocks::from_int(number.as_u32() as u64));
+
+		// state_root : 32 bytes → 4 felts
+		felts.extend(unsafe_digest_bytes_to_felts::<Goldilocks>(
+			self.state_root.as_ref().try_into().expect("hash is 32 bytes"),
+		));
+
+		// extrinsics_root : 32 bytes → 4 felts
+		felts.extend(unsafe_digest_bytes_to_felts::<Goldilocks>(
+			self.extrinsics_root.as_ref().try_into().expect("hash is 32 bytes"),
+		));
+
+		// digest – injective encoding
+		felts.extend(injective_bytes_to_felts::<Goldilocks>(&self.digest.encode()));
+
+		let poseidon_hash: [u8; 32] = hash_variable_length(felts);
+		poseidon_hash.into()
+	}
+}
+
```

Comment:
I think we should include these [lines](https://github.com/paritytech/polkadot-sdk/blob/master/substrate/primitives/runtime/src/generic/header.rs#L135-L153) as well. there might be downstream code that calls this function instead of using `Header` trait one.

Reply:
also, then you could have our impl inside this function and simply call it in the trait impl override